docs/code-generation.md

```diff
@@ -0,0 +1,135 @@
+plan. Instead, ASO is a controller that facilitates creation and management of
+various AWS service objects using Kubernetes CRD instances.
+
+kubebuilder is the the recommended upstream tool for generating CRDs and
+controller stub code. It is a Golang binary that creates the scaffolding for
+CRDs and controller Golang code. Unfortunately, it doesn't really handle
+multiple API groups in the same source code repository, and we need that for
+ASO. For example, we will have an s3.aws.com and an iam.aws.com API group with
+CRDs representing those different API objects and interfaces.
+
+That said, much of kubebuilder simply sets up some templates and then runs the
+`controller-gen` binary from the sigs.kubernetes.io/controller-tools repository
+to generate Golang code.
+
+Our final option was to build a hybrid custom code generator that used
+controller-runtime under the hood but allowed us to generate controller stub
+code for multiple API groups and place generated code [in directories][7] that
+represented Golang best practices. This option gives us the flexibility to
+generate the files and content for multiple API groups but still stay within
+the recommended guardrails of the upstream Kubernetes community.
+
+[7]: https://github.com/kubernetes-sigs/kubebuilder/issues/1268
+
+## Hybrid custom+controller-runtime proposal
+
+![Multi-phase approach to code generation for ASO](images/multi-phase-code-generation.png)
+
+This approach uses multiple phases of code generation.
+
+The first code generation phase consumes model information from a canonical
+source of truth about an AWS service and the objects and interfaces that
+service exposes and generates one or more files containing code that exposes
+Golang types for those objects. These "type files" should be annotated with the
+marker and comments that will allow the core code generators and controller-gen
+to do its work.
+
+Once we have generated the type files, we need to generate a number of basic
+scaffolding for consumers of those types. The upstream code-generator project
+contains generators for this scaffolding. We should be able to make a modified
+version of the upstream [`generate-groups.sh`][8] script to generate all the
+defaults, deepcopy, informers, listers and clientset code.
+
+[8]: https://github.com/kubernetes/code-generator/blob/6b257a9d6f461b5e15dc2f0d13e29731a5b5255a/generate-groups.sh
+
+Next, we will need to generate some skeleton code for the reconciling
+controller handling the new API along with the YAML files representing the CRDs
+that will get loaded into kube-apiserver and the YAML files for RBAC and
+OpenAPI v3 schemas for the CRDs. This is where the [`controller-gen`][9] tool
+from the sigs.kubernetes.io/controller-tools project will come in handy.
+
+[9]: https://github.com/kubernetes-sigs/controller-tools/blob/a5fa7b956b85a6e792bc7086fedf7107d62452b1/cmd/controller-gen/main.go
+
+We also want to generate some Golang code that "installs" a new reconciling
```

Comment:
I think we should separate the build and operate (install/upgrade/ALM) concerns.

Reply:
done.